src/core/thread/link_metrics.cpp

```diff
@@ -0,0 +1,278 @@
+{
+    otError                 error = OT_ERROR_NONE;
+    Tlv                     tlv;
+    LinkMetricsQueryId      linkMetricsQueryId;
+    LinkMetricsQueryOptions linkMetricsQueryOptions;
+    uint8_t                 length = 0;
+    static const uint8_t    tlvs[] = {Mle::Tlv::kLinkMetricsReport};
+    uint8_t                 buf[sizeof(Tlv) * 3 + sizeof(LinkMetricsQueryId) + sizeof(LinkMetricsQueryOptions)];
+
+    // Link Metrics Query TLV
+    tlv.SetType(Mle::Tlv::kLinkMetricsQuery);
+    length += sizeof(Tlv);
+
+    // Link Metrics Query ID sub-TLV
+    linkMetricsQueryId.Init();
+    linkMetricsQueryId.SetSeriesId(aSeriesId);
+    memcpy(buf + length, &linkMetricsQueryId, sizeof(Tlv) + linkMetricsQueryId.GetLength());
+    length += sizeof(Tlv) + linkMetricsQueryId.GetLength();
+
+    // Link Metrics Query Options sub-TLV
+    if (aTypeIdFlagsCount > 0)
+    {
+        linkMetricsQueryOptions.Init();
+        linkMetricsQueryOptions.SetLinkMetricsTypeIdList(aTypeIdFlags, aTypeIdFlagsCount);
+
+        memcpy(buf + length, &linkMetricsQueryOptions, sizeof(Tlv) + linkMetricsQueryOptions.GetLength());
+        length += sizeof(Tlv) + linkMetricsQueryOptions.GetLength();
+    }
+
+    // Set Length for Link Metrics Report Tlv
+    tlv.SetLength(length - sizeof(Tlv));
+    memcpy(buf, &tlv, sizeof(Tlv));
+
+    SuccessOrExit(error = Get<Mle::MleRouter>().SendDataRequest(aDestination, tlvs, sizeof(tlvs), 0, buf, length));
+
+exit:
+    return error;
+}
+
+otError LinkMetrics::AppendSingleProbeLinkMetricsReport(Message &                      aMessage,
+                                                        uint8_t &                      aLength,
+                                                        const LinkMetricsQueryOptions *aQueryOptions,
+                                                        const int8_t                   aNoiseFloor,
+                                                        const Message &                aRequestMessage)
+{
+    otError                  error = OT_ERROR_NONE;
+    LinkMetricsReportSubTlv  metric;
+    const LinkMetricsTypeId *linkMetricsTypeIds;
+    uint8_t                  metricsCount = 0;
+
+    OT_ASSERT(aQueryOptions != nullptr);
+    metricsCount       = kLinkMetricsMaxTypeIdFlags;
+    linkMetricsTypeIds = aQueryOptions->GetLinkMetricsTypeIdList(metricsCount);
+
+    for (uint8_t i = 0; i < metricsCount; i++)
+    {
+        // Link Metrics Report sub-TLVs
+        metric.Init();
+        metric.SetMetricsTypeId(linkMetricsTypeIds[i]);
+
+        switch (linkMetricsTypeIds[i].GetMetricsId())
+        {
+        case OT_LINK_METRICS_PDU_COUNT:
+            if (linkMetricsTypeIds[i].IsLengthFlagSet())
+            {
+                metric.SetMetricsValue32(aRequestMessage.GetPsduCount());
+            }
+            else
+            {
+                metric.SetMetricsValue8(aRequestMessage.GetPsduCount());
+            }
```

Comment:
Does the `SetMetricsValue32`make any sense if the return value of `aRequestMessage.GetPsduCount()` is of type `uint8_t`?

https://github.com/openthread/openthread/blob/2a154be24942f628c094874ea2ffd083ef38cfbd/src/core/common/message.hpp#L855-L861

Reply:
I think in single probe, in no ways the count would be more than `255`(The `Data Request` is divided into 255 fragments, seems impossible). But as I understand, the requester could set the `LengthFlag` to ask for a 4-byte value even though it's not meaningful at all. I remember there's such case in test harness.

Reply:
I think the count is the number of packets feeding into the LqiAverager, other than the fragments of `Data Request` (in fact, I think in general `Data Request` wouldn't be fragmented).

Reply:
It was discussed in JIRA(SPEC-890) that the count in single probe is the count of frames for that message.